homeassistant/components/bang_olufsen/media_player.py

```diff
@@ -358,6 +387,82 @@ def _async_update_volume(self, data: VolumeState) -> None:
 
         self.async_write_ha_state()
 
+    @callback
+    async def _async_update_beolink(self) -> None:
+        """Update the current Beolink leader, listeners, peers and self."""
+
+        # Add Beolink listeners / leader
+
+        self._remote_leader = self._playback_metadata.remote_leader
+
+        # Create group members list
+        group_members = []
+
+        # If the device is a listener.
+        if self._remote_leader is not None:
+            # Add leader
+            group_members.append(
+                cast(str, self._get_entity_id_from_jid(self._remote_leader.jid))
+            )
+
+            # Add self
+            group_members.append(
+                cast(str, self._get_entity_id_from_jid(self._beolink_jid))
+            )
+
+        # If not listener, check if leader.
+        else:
+            beolink_listeners = await self._client.get_beolink_listeners()
+
+            # Check if the device is a leader.
+            if len(beolink_listeners) > 0:
+                # Add self
+                group_members.append(
+                    cast(str, self._get_entity_id_from_jid(self._beolink_jid))
+                )
+
+                # Get the friendly names for the listeners from the peers
+                group_members.extend(
+                    [
+                        cast(str, self._get_entity_id_from_jid(beolink_listener.jid))
+                        for beolink_listener in beolink_listeners
+                    ]
+                )
+
+        self._attr_group_members = group_members
+
+        self.async_write_ha_state()
+
+    def _get_entity_id_from_jid(self, jid: str) -> str | None:
+        """Get entity_id from Beolink JID (if available)."""
+
+        unique_id = jid.split(".")[2].split("@")[0]
+
+        entity_registry = er.async_get(self.hass)
+        return entity_registry.async_get_entity_id(
+            Platform.MEDIA_PLAYER, DOMAIN, unique_id
+        )
+
+    def _get_beolink_jid(self, entity_id: str) -> str | None:
+        """Get beolink JID from entity_id."""
+        jid = None
+
+        entity_registry = er.async_get(self.hass)
+
+        entity_entry = entity_registry.async_get(entity_id)
+        if entity_entry:
+            config_entry = cast(
+                ConfigEntry,
+                self.hass.config_entries.async_get_entry(
+                    cast(str, entity_entry.config_entry_id)
+                ),
+            )
+
+            with contextlib.suppress(KeyError):
+                jid = cast(str, config_entry.data[CONF_BEOLINK_JID])
```

Comment:
Isn't this a roundabout way to find the JID, can't it be reconstructed from the entity's `unique_id` which is the serial number?

Reply:
No, the JID is unfortunately made up of an type number, item number and serial number, as can be seen in the zeroconf discovery  step:

```python
self._beolink_jid = f"{discovery_info.properties[ATTR_TYPE_NUMBER]}.{discovery_info.properties[ATTR_ITEM_NUMBER]}.{self._serial_number}@products.bang-olufsen.com"

```

Reply:
OK, I see. In that case, we need to clean up the error handling instead of just `cast` everything, there's no guarantee the user has passed an `entity_id` which is that of a B&O media player entity.

Also, why is the type checker allowing shenanigans like this, shouldn't it fail because the type if jid is changing from `None` to `str`?
```py
jid = None
...
jid = cast(str, config_entry.data[CONF_BEOLINK_JID])
```

Reply:
OK. I've added a bunch of checks, so the output should be more predictable.